### PR TITLE
[IMP] web_tour: improve tour tip responsiveness and visibility

### PR DESCRIPTION
**Pull request by bso-odoo**
Because it was sometimes difficult to notice the tour tip, added a border to the tour tip icon and made it displayed earlier.

Description of the issue/feature this PR addresses: 
task-2357170

Current behavior before PR:
- the tour tip icon had no extra border
- tour tip display was delayed by an additional 750ms whenever some mutation happened within the DOM (e.g. clicking on the BLOCK header of the options panel contiuously prevented the tip from ever being displayed)

Desired behavior after PR is merged:
- the tour tip icon has an extra border that fades between light and dark gray to make it visible on any background
- DOM mutations are grouped in a first time interval of 750ms, further bursts within the next 750ms are triggered right away but prevent any new triggering within the new 750ms

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_tour/static/src/js/tour_service.js [hunk 1]
```diff
@@ -116,25 +117,34 @@ return session.is_bound.then(function () {
...
-            clearTimeout(mutationTimer);
             currentMutations.push(...mutations);
-            mutationTimer = setTimeout(() => _processMutations(), 750);
+            if (!mutationTimer) { // no timer already set
+                const now = new Date().getTime();
+                const delay = now >= nextProcessMutation ? duration : nextProcessMutation - now;
+                // until time is reached (or after minimal duration, if new burst)
+                mutationTimer = setTimeout(() => _processMutations(), delay);
+            }
         });
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/js/tour_service.js:148]: I think this implementation is the same as doing this:

```javascript
const throttledProcessMutations = _.throttle(_processMutations, duration, {leading: false});
const observer = new MutationObserver(mutations => {
    currentMutations.push(...mutations);
    throttledProcessMutations();
});
```

Could you verify?

I am still unsure about only throttling with a "small" delay though, but we can test on a multi build later on to see what's what. After all, this might just make our test system more strict I guess (forcing the DOM to be stable after 750ms after any change...) but the goal is not to have a runbot red half the time in stable versions either ;)

Comment (bso-odoo) [reply addons/web_tour/static/src/js/tour_service.js]: Equivalence confirmed, implementation adjusted

Comment (qsm-odoo) [reply addons/web_tour/static/src/js/tour_service.js]: @bso-odoo Could you review your git history? I think you squashed your implementation with the `_.throttle` in the wrong commit. Your second commit is still the old implementation

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_tour/static/src/scss/keyframes.scss [hunk 2]
```diff
@@ -47,3 +47,36 @@
         transform: translate(0, 3px);
     }
 }
+
+$o-tip-border-width: 2px; // from tip.scss
+$o-tip-border-animated-width: 8px;
+$o-tip-border-width-correction: $o-tip-border-animated-width - $o-tip-border-width;
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/scss/keyframes.scss:53]: You are forced to do that because you are animating a border-width, which is very inefficient as it forces a whole document paint refresh. Animate something efficient which does not impact the tip layout: I think animating the box-shadow (which can be multiple) works on all supported browser. If not there are other solution to make it efficient: like creating a new element to simulate that border and animate it via transform scale, which is efficient too.

File: addons/web_tour/static/src/js/tip.js [hunk 3]
```diff
@@ -385,8 +385,12 @@ var Tip = Widget.extend({
...
     _unbind_anchor_events: function () {
         this.$anchor.off(".anchor");
-        this.$consumeEventAnchors.off(".anchor");
-        this.$scrolableElement.off('.Tip');
+        if (this.$consumeEventAnchors) {
+            this.$consumeEventAnchors.off(".anchor");
+        }
+        if (this.$scrolableElement) {
+            this.$scrolableElement.off('.Tip');
+        }
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/js/tip.js:393]: Why is this necessary? You found cases where the tip are destroyed while not fully initialized?

File: addons/web_tour/static/src/js/tour_manager.js [hunk 4]
```diff
@@ -87,6 +87,21 @@ return core.Class.extend(mixins.EventDispatcherMixin, ServicesMixin, {
...
+                    case "done":
+                        return true;
+                    default:
+                        tour.beyondReadyState = "started";
+                        setTimeout(function() {
+                            tour.beyondReadyState = "done";
+                        }, delay || 1000);
+                        return false;
+                }
+            },
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/js/tour_manager.js:104]: Sorry I don't understand the point of this commit... you are speaking of "the test cursor was closed right after the
last step of the tour was completed" but this commit seems to change the opposite by delaying the first step of the tour... which is already the point of the "wait_for" property.

Can you re-explain it to me?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

**Comment by qsm-odoo** [comment]
@d-fence We are making modifications to the tour system here, would it be possible to have a multi-build config for the `14.0-tourtip-bso` branch? Thank you! :+1:

**Comment by d-fence** [comment]
> @d-fence We are making modifications to the tour system here, would it be possible to have a multi-build config for the `14.0-tourtip-bso` branch? Thank you!

Yes Sir. Only with `:TestWEventRegister` tests I suppose.

**Comment by qsm-odoo** [comment]
@d-fence no, we are potentially impacting all tests (the relevant commit here is the second one) :wink:

File: addons/web_tour/static/src/scss/keyframes.scss [hunk 5]
```diff
@@ -47,3 +47,36 @@
...
+$o-tip-border-color-light: #F9F9F9; // Odoo brand asset/Light gray
+$o-tip-border-color-dark: #8F8F8F; // Odoo brand asset/Gray
```
[the remark this answers is not in the document]
Comment (bso-odoo) [reply addons/web_tour/static/src/scss/keyframes.scss:55]: Not per say... actual values appear in some elements, but when assigned to variables they look like brand-related info which might be overridden in a theme - which might lead to non-contrasting colors in some configurations.
I switched to white & gray because they are already used within the tip component.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_tour/static/src/scss/keyframes.scss]: I meant using a variable from /web/primary_variables.scss which are not dependent on any theme, but indeed I don't think there is a good candidate there. Keep your hardcoded values :+1:

File: addons/web_tour/static/src/js/tour_service.js (hunk 1 above)
Comment (bso-odoo) [reply addons/web_tour/static/src/js/tour_service.js]: It was actually not equivalent: throttle will not take into account the last set of throttled events - reverting the previous approach

Comment (qsm-odoo) [reply addons/web_tour/static/src/js/tour_service.js:148]: >  throttle will not take into account the last set of throttled events

What do you mean?

Comment (bso-odoo) [reply addons/web_tour/static/src/js/tour_service.js:148]: Beyond the first throttled group, if new events come in within the time interval, they will be collected in the mutations list, but the throttled method will not trigger. If no new event comes in after the time interval, the throttled method will not be called anymore, and the collected mutations will keep waiting in their list without ever being accessed.

**Comment by bso-odoo** [comment]
Not happy about the TestWEventRegister fix, but the problem also occurs from time to time in 14.0 before changes from this branch. This branch however did increase the frequency at which the problem occurred.
Another ticket should be created to address the root cause of the problem - I personally am running out of ideas on this one. Someone else might spot it.

File: addons/web_tour/static/src/scss/tip.scss [hunk 6]
```diff
@@ -114,11 +140,13 @@
 
     &::after {
         content: "";
-        @include o-position-absolute(-$o-tip-border-width, -$o-tip-border-width, -$o-tip-border-width, -$o-tip-border-width);
-        border-color: $o-tip-color;
-        border-radius: 50%;
-        transition: margin $o-tip-size-duration-out ease $o-tip-size-delay-out, border-radius $o-tip-duration-out linear 0s;
-        background: radial-gradient(lighten($o-tip-color, 7%), $o-tip-color);
+        :not(.o_animated) {
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/scss/tip.scss:143]: That css is just not valid so it cannot be ready for review... an "::after" element cannot contain anything.

File: addons/web_tour/static/src/scss/tip.scss [hunk 7]
```diff
@@ -76,6 +76,32 @@
...
+        }
+        &.top {
+            background-image: url('/web_tour/static/src/img/tip_arrow_down.svg');
+        }
+        &.right {
+            background-image: url('/web_tour/static/src/img/tip_arrow_left.svg');
+        }
+        &.left {
+            background-image: url('/web_tour/static/src/img/tip_arrow_right.svg');
+        }
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/scss/tip.scss:99]: 4 svg for 4 directions is just useless, if we make a svg we will just have to rotate it in css for those cases.

File: addons/web_tour/static/src/scss/tip.scss [hunk 8]
```diff
@@ -76,6 +76,32 @@
...
         }
+        box-shadow: none;
+        border-color: transparent !important;
+        border-width: $o-tip-size / 2 !important;
+        background-color: transparent !important;
+        background-size: $o-tip-size * 2;
+        background-repeat: no-repeat;
+        background-position-x: 50%;
+        background-position-y: 50%;
+
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/scss/tip.scss:87]: Using a SVG had two goals:
- Doing anything we want with a cool animation (we will have to ask a designer for that, the point is also removing the css animation and make it part of the svg) and it has to work when the tip is opened and closed (which is not the case here)
- Removing all the useless css: your commit is adding css so it's counterproductive

Let's pause this PR for now. I don't think we will have a viable SVG solution for 14.0 right now. I'll think about it tomorrow :wink:

File: addons/web_tour/static/src/js/tip.js [hunk 9]
```diff
@@ -110,6 +111,7 @@ var Tip = Widget.extend({
         return this._super.apply(this, arguments);
     },
     destroy: function () {
+        this.destroyed = true;
```
Comment (qsm-odoo) [inline addons/web_tour/static/src/js/tip.js:114]: A function `isDestroyed` exists

**Review by qsm-odoo — COMMENTED** [review]
Some comments but don't touch this PR anymore, I'll investigate and test tomorrow to see how this can move forward in stable versions :+1:

File: addons/web_tour/static/src/scss/tip.scss (hunk 7 above)
Comment (bso-odoo) [reply addons/web_tour/static/src/scss/tip.scss:99]: I did it this way thinking that a designer might deliver distinct images - e.g. for some light effects that would depend on the orientation

**Comment by bso-odoo** [comment]
This is obsolete. It was solved totally differently in commit f8df10dd01a969abb861e8eb3ba021d627750d89.
